### ISSUE TITLE
flow on occluded areas

### ISSUE THREAD

**Issue by maorp** [princeton-vl/RAFT#57]
Hi, thanks for sharing this nice work! I'm trying to match pixels between rgb aligned depthmaps using the output flow (i have a calibrated stereo system). I wanted to consult about how to handle the flow on occluded areas where its not very accurate. does RAFT has any ability to output an error\uncertainty measure?

**Comment by zachteed** [comment]
The current model does not provide any measure of uncertainty. However, you can estimate the optical flow in both directions, then check for cycle consistency. For non-occluded pixels, there will be cycle consistency, while occluded pixels will cause disagreement in the forward and backward flow.

**Comment by maorp** [comment]
thanks ill try that. why not actually add a loss to the training based on this cycle condition?

**Comment by zisismp4** [comment]
Hi! @maorp did you actually try the cycle consistency? Did it work ?

**Comment by zisismp4** [comment]
@zachteed Can you please explain how the cycle consistency will be implemented?

**Comment by maorp** [comment]
Hi, asZach suggested I implemented it by calling raft again in reverse on
the images. then you can sum up the flow vectors and check its within some
offset threshold. you need to be a bit cautious because you actually need
to use the forward flow to propagate to a certain coordinate, sample the
reverse flow at this coordinate and see if it takes you back to the start
coordinate or close to it.

**Comment by zisismp4** [comment]
Many thanks for your answer @maorp! This is exactly what confuses me, how to " _sample the
reverse flow at this coordinate and see if it takes you back to the start
coordinate or close to it_".
Can you please share share your implementation?

**Comment by maorp** [comment]
I wrote it in matlab. Ill upload soon. In the meantime check
https://github.com/princeton-vl/RAFT/issues/64#issuecomment-748897559. Its
very similar to mine and you can use parts to do the test. You can even use
it to transform one image to the next and test for texture consistency.

**Comment by zisismp4** [comment]
Please tell me your thoughts about what I tried:
I warped frame_2, using the warping function in issue #64, which gave me a prediction for frame_1.
Then, I warped the prediction with the reverse flow and compared the result with the original frame_2.
Is this the correct way to check for the cycle consistency?

```
flow_low1, flow_up1 = model(image1, image2, iters=20, test_mode=True)
flow_low2, flow_up2 = model(image2, image1, iters=20, test_mode=True)

warped_image = warp(image2, flow_up1)
warped_image = warp(warped_image, flow_up2)
```

**Comment by zachteed** [comment]
Sorry kind of late but this code implements forward / backward consistency. You can just replace the original demo code with this function

```
from utils.utils import InputPadder, coords_grid, bilinear_sampler

def demo(args, thresh=1.0):
    model = torch.nn.DataParallel(RAFT(args))
    model.load_state_dict(torch.load(args.model))

    model = model.module
    model.to(DEVICE)
    model.eval()

    with torch.no_grad():
        images = glob.glob(os.path.join(args.path, '*.png')) + \
                 glob.glob(os.path.join(args.path, '*.jpg'))
        
        images = sorted(images)
        for imfile1, imfile2 in zip(images[:-1], images[1:]):
            image1 = load_image(imfile1)
            image2 = load_image(imfile2)

            padder = InputPadder(image1.shape)
            image1, image2 = padder.pad(image1, image2)

            _, flow12 = model(image1, image2, iters=24, test_mode=True)
            _, flow21 = model(image2, image1, iters=24, test_mode=True)

            coords0 = coords_grid(1, image1.shape[2], image1.shape[3]).cuda()
            coords1 = coords0 + flow12
            coords2 = coords1 + bilinear_sampler(flow21, coords1.permute(0,2,3,1))

            err = (coords0 - coords2).norm(dim=1)
            occ = (err[0] > thresh).float().cpu().numpy()

            cv2.imshow('occ', occ)
            cv2.waitKey()
```

**Comment by tnp23** [comment]
Isn't the err for consistency between coords1 and coords 2, not coords0 and coords2?

**Comment by Geniussh** [comment]
> Isn't the err for consistency between coords1 and coords 2, not coords0 and coords2?

No you warp image1 with flow12 to get a warped image2, and "unwarp" your warped image2 with flow21 to get a new image1. You compare both image1, which corresponds to coords0 and coords2.
